A script for exporting tarballs.

trixirt opened the pull request:
Linux distributions have been using the individual ROCm projects release tarballs to package ROCm into their distributions.  To minimize the disruption to that workfow, produce similar tarballs for the rocm-libraries.

This script comes from the llvm-project another monorepo that collected related compiler projects.  The output has been adjusted to be closer to the the older ROCm project tarballs.

https://github.com/llvm/llvm-project/blob/llvmorg-21.1.5/llvm/utils/release/export.sh

## Motivation

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Technical Details

<!-- Explain the changes along with any relevant GitHub links. -->

## Test Plan

<!-- Explain any relevant testing done to verify this PR. -->

## Test Result

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

stellaraccident reviewed (approved):
If this works for your packaging case, we can add it to the pipelines. We should probably do an equivalent one for rocm-systems.

amd-isparry commented on utils/release/export.sh:
```diff
@@ -0,0 +1,188 @@
+
+set -e
+
+projects="composablekernel hipblas hipblas-common hipblaslt hipcub hipdnn hipfft hiprand hipsolver hipsparse hipsparselt miopen rocblas rocfft rocprim rocrand rocsolver rocsparse rocthrust"
```
Use arrays rather than strings with spaces, e.g.

projects=( composablekernel hipblas
   hipblas-common 
)

keep line lengths reasonable. Consider sorting the list alphabetically.

trixirt replied:
Looking carefully at this, I realized we do not want to have to dink this script for every addition or deletion of a project.  Most of the changes are to figure this out programmatically.  See the update.

amd-isparry commented on utils/release/export.sh:
```diff
@@ -0,0 +1,188 @@
+The following list shows the filenames (with <placeholders>) for the artifacts
+that are being generated (given that you don't touch --template).
+
+$(echo "$projects "| sed 's/\([a-z-]\+\) /  * \1-<RELEASE><RC>.tar.xz \n/g')
```
When you have arrays you can just use printf, a builtin, rather than external commands e.g.

printf  '  * %s-<RELEASE><RC>.tar.xz \n' "${projects[@]}"

consider carefully if you really want this inside a HERE document, or as a separate command.

trixirt replied:
I took this logic from llvm and atm it provides what the possible output would look like.  The generation of the lists is easier as strings with spaces.

amd-isparry commented on utils/release/export.sh:
```diff
@@ -0,0 +1,188 @@
+
+In order to generate snapshots of the upstream main branch you could do this for example:
+
+  $(basename $0) --git-ref develop --template '\${PROJECT}-\${YYYYMMDD}.tar.xz'
```
I would replace the $(basename $0) with just $0. If you need to specify a particular file to print the message you will need to do the same to run the command.

amd-isparry commented on utils/release/export.sh:
```diff
@@ -0,0 +1,188 @@
+export_sources() {
+    local tag="rocm-$release"
+
+    src_dir=$(readlink -f $(dirname "$(readlink -f "$0")")/../..)
```
At least have a comment on why the outer readlink is needed given you have canonicalized the path with the inner one.

Icarus's rule of thumb is that every dollar sign in a shell script should be inside a double quote unless there is a reason why it mustn't be. At the least the $(dirname ...) should be wrapped in double quotes.

trixirt replied:
I changed this from the llvm to something closer to my personal tastes. the readlink is replaces with a realpath.

amd-isparry commented on utils/release/export.sh:
```diff
@@ -0,0 +1,188 @@
+    local tag="rocm-$release"
+
+    src_dir=$(readlink -f $(dirname "$(readlink -f "$0")")/../..)
+    [ -d $src_dir/.git ] || ( echo "No git repository at $src_dir" ; exit 1 )
```
This ignores things like GIT_DIR where you can have the files in different trees to the repo itself.

trixirt replied:
Does the current mono repo do this ?

amd-isparry commented on utils/release/export.sh:
```diff
@@ -0,0 +1,188 @@
+
+    if [ -n "$snapshot" ]; then
+	pushd $src_dir
+	cdate=`git log -1 --date=format:%Y%m%d --format=%cd $snapshot`
```
Use $(...) rather than `...` for consistency with the rest of the script.

amd-isparry commented on utils/release/export.sh:
```diff
@@ -0,0 +1,188 @@
+
+Usage: $(basename $0) [-release|--release <major>.<minor>.<patch>]
+                      [-rc|--rc <num>]
+                      [-final|--final]
```
Why have a --final flag at all, if you don't want an rc then don't give the rc a value.

amd-isparry commented on utils/release/export.sh:
```diff
@@ -0,0 +1,188 @@
+    
+    tag="rocm-$release"
+
+    if [ "$rc" = "final" ]; then
```
The rc logic is convoluted. It can have essentially 3 values, the empty string, the magic word final and anything else as long as it starts with rc..  If it is the empty string then tag gets a dash appended.

trixirt replied:
rc at this point is wishful thinking. whereas the llvm-project does have rc's, rocm-libraries does not.  I could remove the rc logic or change it to suit what we want to do for rc's

amd-isparry commented on utils/release/export.sh:
```diff
@@ -0,0 +1,188 @@
+        echo "Creating tarball for $proj ..."
+        pushd $src_dir/shared/$proj
+        git archive --prefix=$proj-$release$rc/ $tree_id . | xz -T0 >$target_dir/$(template_file $proj)
+        popd
```
Using the "-C" flag to git simplifies this to

        git -C "$src_dir/shared/$proj" archive --prefix=$proj-$release$rc/ $tree_id . | xz -T0 >$target_dir/$(template_file $proj)

Consider if you want to force the use of xz, for example zstd is typically smaller output and less resource intensive to produce.

trixirt replied:
this logic changed alot. I am using xz because that is what llvm uses so the code can be reused.

amd-isparry commented on utils/release/export.sh:
```diff
@@ -0,0 +1,188 @@
+
+while [ $# -gt 0 ]; do
+    case $1 in
+        -release | --release )
```
As a matter of convenience, I like to put in the optional open bracket, it makes it easier for editors which don't understand shell syntax to at least show matches. It also avoids some very obscure corner cases.

amd-isparry commented on utils/release/export.sh:
```diff
@@ -0,0 +1,188 @@
+            ;;
+        -rc | --rc )
+            shift
+            rc="rc$1"
```
Adding the "rc" seems restrictive.

amd-isparry commented on utils/release/export.sh:
```diff
@@ -0,0 +1,188 @@
+}
+
+while [ $# -gt 0 ]; do
+    case $1 in
```
You could change a leading "--" to "-" and remove the need for all the duplicated words in the case statements, although it would then recognize "--h" as a request for help.

   case "${1/#--/-}" in
   ( "-release" ) shift; release="$1" ;;
   ( "-rc" ) shift ; rc="rc$1" ;;

trixirt replied:
done.

amd-isparry reviewed (commented):
Running "shellcheck" on this will identify more potential issues.

trixirt committed:
A script for exporting tarballs.

Linux distributions have been using the individual ROCm projects release
tarballs to package ROCm into their distributions.  To minimize the disruption
to that workfow, produce similar tarballs for the rocm-libraries.

This script comes from the llvm-project another monorepo that collected
related compiler projects.  The output has been adjusted to be closer to the
the older ROCm project tarballs.

https://github.com/llvm/llvm-project/blob/llvmorg-21.1.5/llvm/utils/release/export.sh

Signed-off-by: Tom Rix <Tom.Rix@amd.com>

stellaraccident commented:
Thanks for making it yours. I can't say I love the original but decided to not nitpick since I was familiar with where this came from and figured you'd ultimately end up rewriting the script once you got some mileage on it. 

Ftr - here is the more full featured version for creating a source archive of the whole project: https://github.com/ROCm/TheRock/blob/main/build_tools/export_source_archive.py

(Not saying you should use that for this simpler case)
